chainer/functions/activation/softmax.py

```diff
@@ -64,26 +83,35 @@ def backward(self, x, gy):
             handle = cudnn.get_handle()
             gx = xp.empty_like(y)
             gx_tensor4d = cuda.cupy.ascontiguousarray(
-                gx.reshape(self._get_tensor4d_shape(gx.shape)))
-            gy = cuda.cupy.ascontiguousarray(gy[0])
+                gx.reshape(_get_tensor4d_shape(self.axis, gx.shape)))
+            gy = cuda.cupy.ascontiguousarray(gy)
             desc = cudnn.create_tensor_descriptor(gx_tensor4d)
             libcudnn.softmaxBackward(
                 handle, _algorithm, _mode, one.data, desc.value,
                 y.data.ptr, desc.value, gy.data.ptr, zero.data,
                 desc.value, gx.data.ptr)
         else:
-            gx = y * gy[0]
+            gx = y * gy
             sumdx = gx.sum(axis=self.axis, keepdims=True)
             gx -= y * sumdx
 
         return gx,
 
-    def _get_tensor4d_shape(self, shape):
-        left_shape = numpy.prod(shape[slice(0, self.axis)], dtype=numpy.int)
-        center_shape = shape[self.axis]
-        right_shape = numpy.prod(
-            shape[slice(self.axis + 1, len(shape))], dtype=numpy.int)
-        return left_shape, center_shape, right_shape, 1
+    def backward(self, indexes, grad_outputs):
+        y, gy = self.get_retained_inputs()
+        ggx, = grad_outputs
+        gs = -chainer.functions.sum(ggx * y, axis=self.axis, keepdims=True)
+        ga = ggx + chainer.functions.broadcast_to(gs, gy.shape)
```

Comment:
Sorry, I was ambiguous. I though it's better to write it as follows.
```
gs = chainer.functions.sum(...)
ga = ggx - chainer.functions.broadcast_to(gs, ...)
```
so that negation+add (two functions) is replaced by subtract (one function).

Reply:
Yes, fixed!